fortnite_api/utils.py

```diff
@@ -0,0 +1,184 @@
+"""
+MIT License
+
+Copyright (c) 2019 Luc1412
+
+Permission is hereby granted, free of charge, to any person obtaining a copy
+of this software and associated documentation files (the "Software"), to deal
+in the Software without restriction, including without limitation the rights
+to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+copies of the Software, and to permit persons to whom the Software is
+furnished to do so, subject to the following conditions:
+
+The above copyright notice and this permission notice shall be included in all
+copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+"""
+
+from __future__ import annotations
+
+import datetime
+from typing import Any, Callable, Dict, Hashable, Tuple, TypeVar, Union
+
+try:
+    import orjson  # type: ignore
+
+    _has_orjson: bool = True
+except ImportError:
+    import json
+
+    _has_orjson: bool = False
+
+K_co = TypeVar('K_co', bound='Hashable', covariant=True)
+V_co = TypeVar('V_co', covariant=True)
+T = TypeVar('T')
+
+__all__: Tuple[str, ...] = ('parse_time', 'copy_doc', 'prepend_doc', 'to_json', 'MISSING')
+
+BACKUP_TIMESTAMP: str = '0001-01-01T00:00:00'
+
+
+class _MissingSentinel:
+    __slots__ = ()
+
+    def __eq__(self, other: Any) -> bool:
+        return False
+
+    def __bool__(self) -> bool:
+        return False
+
+    def __hash__(self) -> int:
+        return 0
+
+    def __repr__(self):
+        return '...'
+
+
+MISSING: Any = _MissingSentinel()
+
+
+if _has_orjson:
+
+    def to_json(string: Union[str, bytes]) -> Dict[Any, Any]:
+        return orjson.loads(string)  # type: ignore
+
+else:
+
+    def to_json(string: Union[str, bytes]) -> Dict[Any, Any]:
+        return json.loads(string)  # type: ignore
+
+
+def parse_time(timestamp: str) -> datetime.datetime:
+    # This can happen when the API is supposed to return a timestamp but there is no timestamp to give, so it yields an improper timestamp without a UTC offset.
+    if timestamp == BACKUP_TIMESTAMP:
+        return datetime.datetime.fromisoformat(timestamp)
+
+    # If the timestamp str contains ms or us, strptime with them. If not, fallback
+    # to default strptime.
+    try:
+        return datetime.datetime.strptime(timestamp, '%Y-%m-%dT%H:%M:%S.%f%z')
+    except ValueError:
+        return datetime.datetime.strptime(timestamp, '%Y-%m-%dT%H:%M:%S%z')
+
+
+def now() -> datetime.datetime:
```

Comment:
Not in this case because `utcnow()` returns the UTC date as a naive datetime obj - ie. it doesn't contain any timezone info. It's important to keep the obj aware of timezones for proper ratelimit handling.

Reply:
Sorry, my comment was a bit too broad. I meant renaming the function to `utcnow()`. In 3.12 `utcnow()` is deprecated anyway.